patroni/__init__.py

```diff
@@ -50,19 +55,20 @@ def initialize(self):
 
         # is data directory empty?
         if self.postgresql.data_directory_empty():
-            # racing to initialize
-            if self.ha.dcs.race('/initialize'):
-                self.postgresql.initialize()
-                self.ha.dcs.take_leader()
-                self.postgresql.start()
-                self.postgresql.create_replication_user()
-                self.postgresql.create_connection_users()
-            else:
-                while True:
```

Comment:
I don't see us waiting for the leader in the code fragment you've highlighted. What has changed is that now we try to bootstrap the PostgreSQL cluster before taking the leader: doing it the other way around provides no benefits, but bears an additional responsibility to remove the leader flag during cleanup_on_failed_initialization.

Actually, I see one problem here: if the initialiser node suddenly dies after initidb (say, it cannot not create the users), and then restarts - it will think it was fully initialised and will try to start as a normal member that was started after a stop, potentially acquiring the master role (which would screw up the whole cluster), or a replica, in which case it won't be able to catchup with the master, because it was not initialised from it.